nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.nessie;
+
+import com.dremio.nessie.client.NessieClient;
+import com.dremio.nessie.error.NessieConflictException;
+import com.dremio.nessie.error.NessieNotFoundException;
+import com.dremio.nessie.model.Contents;
+import com.dremio.nessie.model.ContentsKey;
+import com.dremio.nessie.model.IcebergTable;
+import com.dremio.nessie.model.ImmutableIcebergTable;
+import java.util.Map;
+import org.apache.iceberg.BaseMetastoreTableOperations;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.io.FileIO;
+
+/**
+ * Nessie implementation of Iceberg TableOperations.
+ */
+public class NessieTableOperations extends BaseMetastoreTableOperations {
+
+  private final NessieClient client;
+  private final ContentsKey key;
+  private UpdateableReference reference;
+  private IcebergTable table;
+  private FileIO fileIO;
+
+  /**
+   * Create a nessie table operations given a table identifier.
+   */
+  public NessieTableOperations(
+      ContentsKey key,
+      UpdateableReference reference,
+      NessieClient client,
+      FileIO fileIO) {
+    this.key = key;
+    this.reference = reference;
+    this.client = client;
+    this.fileIO = fileIO;
+  }
+
+  @Override
+  protected void doRefresh() {
+    try {
+      reference.refresh();
+    } catch (NessieNotFoundException e) {
+      throw new RuntimeException("Failed to refresh as ref is no longer valid.", e);
```

Comment:
We may want to let the original exception propagate instead of throwing a `RuntimeException` so that people working with the table can take action if the ref is gone. I'm not really sure what the right thing is here.

Reply:
I would normally agree but `NessieNotFoundException` is a checked exception and I don't want to add that throws clause to the interface. Is there a more specialised unchecked exception that you think would suit?

If it helps the likelihood of this throwing should be very low in real life. Typically a user/service would own their branches and deleting a branch when someone else is actively working on it would be an organisational issue rather than something that should be handled in code. An analogy in git would be someone deleting your branch which is an active PR (without closing the PR).